src/PlottingInterface.cc

```diff
@@ -243,10 +292,85 @@ void Topic::Callback(const google::protobuf::Message &_msg)
       data = this->dataPtr->FieldData(_msg, field);
     }
 
+    if (!fieldIt.second)
+        continue;
+
+    // Field Arrival Time
+    fieldIt.second->SetTime(headerTime);
+
+    // Field Value
     fieldIt.second->SetValue(data);
+
+    // Update Field Charts UI
+    this->UpdateGui(fieldIt.first);
   }
 }
 
+//////////////////////////////////////////////////////
+bool Topic::HasHeader(const google::protobuf::Message &_msg,
+                      double &_headerTime)
+{
+  auto ref = _msg.GetReflection();
+  auto header = _msg.GetDescriptor()->FindFieldByName("header");
+  auto found = ref->HasField(_msg, header);
+
+  if (!found)
+    return false;
+
+  auto stamp = header->message_type()->FindFieldByName("stamp");
+
+  if (!stamp)
+      return false;
+
+  auto headerMsg = ref->MutableMessage
+          (const_cast<google::protobuf::Message *>(&_msg), header);
+
+  if (!headerMsg)
+      return false;
+
+  ref = headerMsg->GetReflection();
+
+  auto stampMsg = ref->MutableMessage
+          (const_cast<google::protobuf::Message *>(headerMsg), stamp);
+
+  if (!stampMsg)
+      return false;
+
+  auto secField = stamp->message_type()->FindFieldByName("sec");
+  auto nsecField = stamp->message_type()->FindFieldByName("nsec");
+
+  auto sec = this->dataPtr->FieldData(*stampMsg, secField);
+  auto nsec = this->dataPtr->FieldData(*stampMsg, nsecField);
+
+  _headerTime = sec + nsec * std::pow(10, -9);
+
+  return true;
+}
+
+//////////////////////////////////////////////////////
+void Topic::UpdateGui(const std::string &_field)
+{
+  auto field = this->dataPtr->fields[_field];
+
+  auto x = field->Time();
+  auto y = field->Value();
+
+  QString fieldFullPath = QString::fromStdString
+          (this->dataPtr->name + "-" + _field);
+
+  auto charts = field->Charts();
+
+  for (auto const &chart : charts)
+    emit plot(chart, fieldFullPath, x, y);
+}
+
+//////////////////////////////////////////////////////
+void Topic::SetPlottingTimeRef(const double *_timePtr)
+{
+  if (!this->dataPtr->plottingTime)
```

Comment:
I think callers of this function would expect that whenever you set the reference, the pre-existing reference is updated. Is this check needed?